File: App02_Library/app/src/main/java/com/wooseok/library/MainActivity.java
package com.wooseok.library;

import android.os.Bundle;

import com.google.android.material.snackbar.Snackbar;

import androidx.appcompat.app.AppCompatActivity;

import android.view.View;

import androidx.navigation.NavController;
import androidx.navigation.Navigation;
import androidx.navigation.ui.AppBarConfiguration;
import androidx.navigation.ui.NavigationUI;

import com.wooseok.library.databinding.ActivityMainBinding;

import android.view.Menu;
import android.view.MenuItem;

public class MainActivity extends AppCompatActivity {

    private AppBarConfiguration appBarConfiguration;

    /**
     * viewbinding이 true로 설정이 되면
     * 자동으로 내부에서 생성되는 새로운 개념의 클래스이다
     */
    private ActivityMainBinding binding;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        /**
         * setContentView(R.id.activity_main) 형식으로 *.xml 파일을 열여서
         * 화면을 구성했던 코드가 다음과 같이 변경된다
         */
        binding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        /**
         * binding이 성립되면 *.xml에 포함된 View Component는 별도의
         * 객체 선언이 필요 없고
         * 사용할때는
         *      binding.txt_name
         *      binding.txt_msg
         *  와 같은 형식으로 직접 핸들링이 가능하다
         */

        setSupportActionBar(binding.toolbar);

        // ActionBar에 설정된 햄버거 버튼과 관련 코드
        NavController navController = Navigation.findNavController(this, R.id.nav_host_fragment_content_main);
        appBarConfiguration = new AppBarConfiguration.Builder(navController.getGraph()).build();
        NavigationUI.setupActionBarWithNavController(this, navController, appBarConfiguration);

        // fab = findByViewId(R.id.fab) 이런 코드가 필요했으나
        // binding이 성립된 경우는 필요 없는 코드가 된다.
        // binding.fab 객체에 접근하여 event등을 선언할 수 있게 된다.

        binding.fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Snackbar.make(view, "Replace with your own action", Snackbar.LENGTH_LONG)
                        .setAction("Action", null).show();
            }
        });
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();

        //noinspection SimplifiableIfStatement
        if (id == R.id.action_settings) {
            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    public boolean onSupportNavigateUp() {
        NavController navController = Navigation.findNavController(this, R.id.nav_host_fragment_content_main);
        return NavigationUI.navigateUp(navController, appBarConfiguration)
                || super.onSupportNavigateUp();
    }
}

File: HelloApp2/app/src/main/java/com/wooseok/hello/MainActivity.kt
package com.wooseok.hello

import android.os.Bundle
import com.google.android.material.snackbar.Snackbar
import androidx.appcompat.app.AppCompatActivity
import androidx.navigation.findNavController
import androidx.navigation.ui.AppBarConfiguration
import androidx.navigation.ui.navigateUp
import androidx.navigation.ui.setupActionBarWithNavController
import android.view.Menu
import android.view.MenuItem
import com.wooseok.hello.databinding.ActivityMainBinding

/**
 * Android App 에서 화면 View 와 관련된 클래스
 * main() 진입점 클래스
 */
class MainActivity : AppCompatActivity() {

    /**
     * NoActionBar Theme 화면에서
     * Custom 한 ActionBar(title bar)를 생성하기 위한 클래스
     */
    private lateinit var appBarConfiguration: AppBarConfiguration

    /**
     * res/layout/activity_main.xml 파일을 클래스코드에서 핸들링을 할때
     * xml 파일을 사용하여 자동으로 클래스가 생성된다
     */
    private lateinit var binding: ActivityMainBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        /**
         * activity_main.xml 의 view Component 를 한꺼번에
         * 바인딩하기 위한 코드
         * 예) 과거의 Java 코드에서 xml 에 포함된 view Component 들을
         * 사용하기 위하여 초기화 했던 다음과 같은 코드를 더이상 사용하지 않는다. 
         * TextEdit textEdit = findById(R.id.txt_01)
         */
        binding = ActivityMainBinding.inflate(layoutInflater)

        /**
         * activity_main.xml 전체를 확장하여
         * 처음 어플 화면에 보여라
         */
        setContentView(binding.root)

        /**
         * 임의로 생성한 ActionBar 를 현재 화면에 같이 부착하라
         * id 가 toolbar 인 ActionBar 컴포넌트를 현재 화면에
         * 부착하라
         */
        setSupportActionBar(binding.toolbar)

        val navController = findNavController(R.id.nav_host_fragment_content_main)
        appBarConfiguration = AppBarConfiguration(navController.graph)
        setupActionBarWithNavController(navController, appBarConfiguration)

        binding.fab.setOnClickListener { view ->
            Snackbar.make(view, "Replace with your own action", Snackbar.LENGTH_LONG)
                    .setAction("Action", null).show()
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        // Inflate the menu; this adds items to the action bar if it is present.
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        return when (item.itemId) {
            R.id.action_settings -> true
            else -> super.onOptionsItemSelected(item)
        }
    }

    override fun onSupportNavigateUp(): Boolean {
        val navController = findNavController(R.id.nav_host_fragment_content_main)
        return navController.navigateUp(appBarConfiguration)
                || super.onSupportNavigateUp()
    }
}